### DISCUSSION TITLE
Is it possible to write a serverside app using AS3/AIR?

### DISCUSSION THREAD

**Discussion by GeneralVimes** [airsdk/Adobe-Runtime-Support#2078] [Q&A]
Basically, that’s the question. I have an idea of a client-server app and I wonder, will it be possible to code the serverside on AS3?

**Comment by ajwfrost** [comment]
Why not? We had a simple server-side app we wrote a while ago where we just used AIR (ServerSocket and a really basic http engine, although there are open source options available for more complete http handling) to trigger some tools to run locally and return the resulting files via downloads - it ran with a UI that showed basic diagnostic information, but you can always run AIR as a command-line app if you don't use any Stage/DisplayObject stuff.

I'd probably say that this isn't what it's designed for, so you would find better capabilities with other technologies, but depending on what you're after it should be possible to achieve with an AS3 app running in AIR.

Would be interested in what other people have done here.. :-)

thanks

**Comment by Dextercitox** [comment]
Some time ago I developed one online app for Heineken where you generate your favorite festival "poster" with your favorite rock bands. We generate a PDF based on some params and design chooses selected by the users. 

We used a windows server, and develop a main Air app that works as a load-balancer, that communicates with other 4 air apps that generates the posters and save them. The main app receive the request and based on the queue and the load, it redirects with sockets to one of the 4 other instances that generates the PDF.

We choose AIR because it was easy to generate the PDF, that uses different Typography ( and font color, size that changes depending the text ... ), backgrounds, and some graphic stuff that needs some seconds to generate the PDF, and we receive more than 3 or 4 request at the time in busy hours. It was a big pain to made it with PHP because of the rendering of the typography and we also have the advantage of control the queue and load with the main Air APP. I remember to be connected remotely to the Windows Server and see how the main APP runs ( we made a very simple interface ) and was a very funny project. I still have the code over there...

It worked like a charm, it was 10 years ago ! :P

https://noergia.com/wp-content/uploads/2012/09/myfestival1.jpg

**Comment by 2jfw** [comment]
I think the **redtamarin** project from [@Zwetan](https://github.com/zwetan) could give you some insight or inspiration on this:
https://github.com/Corsaair/redtamarin

**Reply by marchbold** [reply]
Had a play around with that several years ago, seemed very promising... Would be interesting to get the changes available in AIR :)

**Reply by ajwfrost** [reply]
Yes .. sadly, I believe Zwetan may have been a victim of the initial wave of covid .. he dropped off all forums/discussions, I didn't manage to get hold of him after that. If anyone has any information on him it would be good to find out for sure..

License terms for the redtamarin means that we probably can't pull that functionality into AIR directly, not sure if it's possible to expose a lot of it via an ANE though?

But if there are particular requests for low-level C APIs to be exposed from ActionScript, you can always add a feature request!

thanks

**Reply by marchbold** [reply]
Great! will do. 

I think two of the things that excited me most about it were:
- command line app generation (headless), which we sort of have now but probably still needs some tweaks to make it as complete as redtamarin;
- the [`as3shebang` interpreter](https://github.com/Corsaair/as3shebang) which ran as3 as a scripting language on the command line, this one I feel opens up a world of node like server applications, hotloading etc;

**Reply by ajwfrost** [reply]
Wow I'd not come across that `as3shebang` thing before, missed that one... but a quick look shows that it's using avmshell code i.e. there is some sort of AS3 interpreter/compiler built into the avmplus project, and if we built AIR with the `VMCFG_EVAL` configuration, we would also be able to do the same...
https://github.com/adobe/avmplus/blob/master/eval/eval-avmplus.cpp#L103

Will have to look at this :-)

On the command-line side, yes the `redtamarin` is a lot more lightweight whereas our command-line option is still having the same bootstrap code, just ignoring anything visual/sprite-based..

**Comment by htmiel** [comment]
On the subject on AIR client-server, I used this ANE with great success to secure a connection over the standard AIR sockets:
https://github.com/airext/openssl

Both client and server apps were made in AIR (server app running on Windows Server). Currently it's only available for Windows and iOS.

From my tests, a single AIR server app can hold up to 4000 TCP simultaneous connections while using up to 1GB of RAM.

And if you have an 8-core server for example, you can run 8 of these apps (that's 32k clients), with local socket conections to communicate between them :)
